Refactored AddNoodeScene, moved localiaztion to vm

gemdev111 opened the pull request:
Added new field Latency
Updated old error state to a new one with retry possability 
Added LatencyMeasureService
Updated AddNodeSceneViewModel to supports LatencyMeasure
Updated performance of the screen
#96

<table>
  <tr>
    <td><img src="https://github.com/gemwalletcom/gem-ios/assets/171273137/39e92aad-a9df-48d3-b78e-0574e70baaba" style="width: 45%;" /></td>
    <td><img src="https://github.com/gemwalletcom/gem-ios/assets/171273137/2d40d386-97b0-4d39-b28b-37880f402326" style="width: 45%;" /></td>
  </tr>
</table>​

![Simulator Screenshot - iPhone SE (3rd generation) - 2024-07-01 at 17 43 22](https://github.com/gemwalletcom/gem-ios/assets/171273137/04bf3d60-589b-4a6f-b921-11a739c4f5b3)

gemdev111 committed:
Refactored AddNoodeScene, moved localiaztion to vm
Added new field Latency
Updated old error state to a new one with retry possability
Added LatencyMeasureService
Updated AddNodeSceneViewModel to supports LatencyMeasure, updated performance of the screen
Added new eng, rus localizations

gemcoder21 commented on Gem/Nodes/Services/LatencyMeasureService.swift:
```diff
@@ -0,0 +1,57 @@
+        var value: Int {
+            let value: Double
+            switch self {
+            case .fast(let double):
```
```suggestion
            case .slow(let double), .normal(let double), .fast(let double):
```

gemcoder21 commented on Gem/Nodes/Services/LatencyMeasureService.swift:
```diff
@@ -0,0 +1,57 @@
+            return Int(value)
+        }
+
+        var formattedValue: String {
```
service should not have formatter/text data. move to view model

gemcoder21 commented on Gem/Nodes/ViewModels/AddNodeSceneViewModel.swift:
```diff
@@ -63,39 +75,35 @@ extension AddNodeSceneViewModel {
-
         do {
-            let (isSynced, networkId, blockNumber) = try await (isNodeSync, chainId, blockNumber)
+            let (latency, chainId) = try await fetchChainID(service: service)
```
those methods should be all async in parallel, currently it's split in two, twice slower

gemdev111 replied:
The idea why we doing that way - to receive firstly chainId, compare  it with chainId we looking for and if chainId is different, do not  use 2 more unneeded requests. The  latency value we are getting with a chainId also. If we do a 3 async let we are not receive proper latency value and in case of not correct chain id, we do 2 more unneeded requests to a node.

gemcoder21 replied:
it does not matter if we receive a latency or not. it needs to run fast, run 3 fetch requests and parallel and present results to the user

gemcoder21 commented on Gem/Nodes/Services/LatencyMeasureService.swift:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    static func measure<T>(for asyncFunction: @escaping () async throws -> T?) async throws -> (latency: Latency, value: T?) {
```
Did you consider extending SwiftHttpClient to provide request method with latency instead?

gemdev111 replied:
Current implementation of getting ping can be appaielble as "adapter pattern" to any async await request/service/whatever we have. 

If we consider to using it in SwiftHttpClient by extending the functionality of request/response with several methods, we also have to extend our existing api services, if I get it correctly orr import SwiftHHTPClient in our viewmodels, modules - which is a one more external dependecy with a limit usage ( because we need a ping only in few places in gem target )

gemcoder21 commented on Gem/Nodes/Scenes/AddNodeScene.swift:
```diff
@@ -81,40 +81,43 @@ extension AddNodeScene {
-                    subtitle: result.isInSync ? "✅" : "❌",
-                    subtitleStyle: .calloutSecondary
+                    title: model.inSyncTitle,
+                    subtitle: result.isInSync ? "✅" : "❌"
```
`result.isInSync ? "✅" : "❌"` let's move this to view model

gemcoder21 commented on Gem/Nodes/ViewModels/AddNodeSceneViewModel.swift:
```diff
@@ -5,15 +5,16 @@ import Primitives
+    
     let chain: Chain
-    @Published var urlString: String = ""
+    @Published var inputFieldValue: String = ""
```
```suggestion
    @Published var url: String = ""
```

gemdev111 committed:
Code-review adjustments
Moved api result value properties to the vm
Minor adjustements ref to discussion w/ @gemcoder111
Removed retry button & action from ListItemErrorView in AddNodeScene

gemcoder21 committed:
Update localization strings
